spike_multinvi.md

```diff
@@ -0,0 +1,186 @@
+Proposed:
+```
+// abstract base class
+class SatOrbit
+{
+ // Interface
+public:
+    SatOrbit() = default;
+    virtual ~SatOrbit() = default;
+
+    std::vector<sat355::TLE> ReadFromFile(int argc, char* argv[]);
+    std::vector<OrbitalData> CalculateOrbitalData(const std::vector<sat355::TLE>& tleVector);
+    void SortOrbitalVector(std::vector<OrbitalData>& ioOrbitalVector);
+    std::vector<std::vector<OrbitalData>> CreateTrains(const std::vector<OrbitalData>& orbitalVector);
+    void PrintTrains(const std::vector<std::vector<OrbitalData>>& trainVector);
+
+// Types
+protected:
+    using OrbitalDataVector = std::tuple<std::mutex, std::vector<OrbitalData>>;
+    using orbit_iterator = std::vector<OrbitalData>::iterator;
+
+// Implementation
+private:
+    // SatOrbit
+    virtual std::vector<sat355::TLE> OnReadFromFile(int argc, char* argv[]) = 0;
+    virtual void OnCalculateOrbitalData(const std::vector<sat355::TLE>& inTLEVector, OrbitalDataVector& ioDataVector) = 0;
+    virtual void OnSortOrbitalVector(std::vector<OrbitalData>& ioOrbitalVector) = 0;
+    virtual std::vector<std::vector<OrbitalData>> OnCreateTrains(const std::vector<OrbitalData>& orbitalVector) = 0;
+    virtual void OnPrintTrains(const std::vector<std::vector<OrbitalData>>& trainVector) = 0;
+};
+
+
+class SatOrbitSingle : public SatOrbit
+{
+public:
+    SatOrbitSingle() = default;
+    virtual ~SatOrbitSingle() = default;
+
+// Implementation
+private:
+    // SatOrbit
+    std::vector<sat355::TLE> OnReadFromFile(int argc, char* argv[]) override;
+    void OnCalculateOrbitalData(const std::vector<sat355::TLE>& inTLEVector, OrbitalDataVector& ioDataVector) override;
+    void OnSortOrbitalVector(std::vector<OrbitalData>& ioOrbitalVector) override;
+    std::vector<std::vector<OrbitalData>> OnCreateTrains(const std::vector<OrbitalData>& orbitalVector) override;
+    void OnPrintTrains(const std::vector<std::vector<OrbitalData>>& trainVector) override;
+};
+
+
+// Multi thread class: Is-a SatOrbit
+class SatOrbitMulti : public SatOrbitSingle
+{
+// Interface
+public:
+    SatOrbitMulti(std::size_t inNumThreads = 1);
+    ~SatOrbitMulti() override = default;
+
+// Types
+private:
+    using tle_const_iterator = std::vector<sat355::TLE>::const_iterator;
+    using IteratorPairVector = std::vector<std::tuple<orbit_iterator, orbit_iterator>>;
+
+// Implementation
+private:
+    // SatOrbit
+    void OnCalculateOrbitalData(const std::vector<sat355::TLE>& inTLEVector, OrbitalDataVector& ioDataVector) override;
+    void OnSortOrbitalVector(std::vector<OrbitalData>& ioOrbitalVector) override;
+
+    // SatOrbitMulti
+    void OnCalculateOrbitalDataMulti(const tle_const_iterator& tleBegin, const tle_const_iterator& tleEnd, OrbitalDataVector& ioDataVector) override;
+    void OnSortOrbitalVectorMulti(orbit_iterator& inBegin, orbit_iterator& inEnd) override;
+    void OnSortMergeVector(orbit_iterator& ioBegin, orbit_iterator& ioMid, orbit_iterator& ioEnd) override;
```

Comment:
These new private virtuals are not `override`...
Instead they are `virtual`: because they are being newly declared starting with classes of type=`SatOrbitMulti`